GH-706 dependency patch fix for 8.x only

brodybits opened the pull request:
UPDATED:
- reintroduce `xcode` as needed in 8.1.x to avoid breaking plugins such as `branch-cordova-sdk` before next major release, to resolve GH-706 (bug #706)
- ~~emit a message in case someone uses `requireCordovaModule('xcode')` that it is now deprecated, will be removed in the near future, along with the recommended solution~~
- reintroduce some other dependencies to better ensure that any other plugins or applications using `requireCordovaModule` would not be broken by a minor release upgrade
- emit a deprecation warning message in case a plugin or application uses `requireCordovaModule` with a non-Cordova module
- some other updates from PR #707

This fix is NOT WANTED in `master` branch. I think we do not want to support `requireCordovaModule('xcode')` in the next major release.

This fix was tested as follows:
* pushed this change to `bug-706-xcode-hotfix-test` on my fork
* made `bug-706-xcode-hotfix-test` branch on my fork of `cordova-cli` that uses cordova-lib from `git+https://github.com/brodybits/cordova-lib.git#bug-706-xcode-hotfix-test` as dependency
* In <https://github.com/brodybits/cordova-xcode-error> (fork of <https://github.com/Tallyb/cordova-xcode-error> with `config.xml` fix):
  - I used cross-platform `nvs` tool (<https://github.com/jasongin/nvs>) to switch to Node.js 10.10 and do `npm i -g https://github.com/brodybits/cordova-cli#bug-706-xcode-hotfix-test`
  - `cordova platform add ios` outputs the deprecation message, does not fail with the `xcode` error

Resolves #706

janpio commented:
(PR title should describe the change first, then administrative information)

raphinesse commented:
I *think* a minor release would be more fitting to introducing a deprecation notice, but I'm not quite sure. What would you say?

raphinesse commented:
To really adhere to semver, we cannot drop any dependencies or make major version updates because of #689. So here's a list of dependencies that we dropped between 8.0.0 and 8.1.0:
```
dependency-ls@^1.1.1
request@2.79.0
unorm@1.4.1
valid-identifier@0.0.1
xcode@^1.0.0
```
Furthermore, we updated `tar` from `2.2.1` to `^4.4.1`.

brodybits commented:
I just pushed some changes:
* updates from #707 
* squash fixes to support Node.js 4 (unsure if 'use strict' is needed or not)
* fix `const pkg` declaration for Node.js 4 & resolve unused variable in eslint
* remove extra xcode warning that we added, due to the more general deprecation warning from #707 (squash fix)

Coming up are changes to restore the dependencies we dropped in 8.1.0.

janpio commented on src/hooks/Context.js:
```diff
@@ -72,6 +91,10 @@ Context.prototype.requireCordovaModule = function (modulePath) {
+    } else if (/xcode/.test(modulePath)) {
+        events.emit('warn',
+            "Use of requireCordovaModule('xcode') is now deprecated, will be removed in the near future." +
+            " Recommended solution: run `npm install xcode` & use `require('xcode')`");
```
All good.

raphinesse commented:
@brodybits ping me when you think this is ready to review, please.

brodybits commented:
@raphinesse this PR is now ready for review.

janpio commented on package.json:
```diff
@@ -24,6 +24,7 @@
     "cordova-js": "^4.2.2",
     "cordova-serve": "^2.0.0",
     "dep-graph": "1.1.0",
+    "dependency-ls": "^1.1.1",
```
this and all the other dependencies that are readded are not reflected in the PR title and description

brodybits commented:
> It's just a deprecation _notice_, right?

Yes. And I would like to resolve the GH-706 in a patch release.

One more thing is that if cordova-lib and Cordova CLI do not have the same version number there some extra "noise" will show up in `cordova --version` like this:

    8.1.2-dev (cordova-lib@8.1.1-dev)

I would favor publishing this in cordova-lib@8.1.2 to avoid such noise.

janpio commented:
> I would favor publishing this in cordova-lib@8.1.2 to avoid such noise.

That's not how semver works. 
That "noise" is expected and should be there (actually not showing the `cordova-lib` version if they are equal is not optimal).

brodybits commented:
As discussed on the #dev channel on Slack, I just removed the changes to emit deprecation warning messages in case of non-Cordova modules (also removed a test case). This is to avoid the issue with <https://semver.org/#spec-item-7> where a minor version increase is needed in case of the deprecation message. Merging now.

janpio commented:
You didn't actually remove the code before merging.

brodybits commented:
What I removed was cherry-pick of <https://github.com/apache/cordova-lib/pull/707/commits/27a2cfd6c53cc268ae7200b877216aa146243a3f> from PR #707. You should not see any trace of it since I did a rebase with the cherry-pick of <https://github.com/apache/cordova-lib/pull/707/commits/27a2cfd6c53cc268ae7200b877216aa146243a3f>. If I look at <https://github.com/apache/cordova-lib/pull/708/files> the code to emit the deprecation warning is no longer there.

janpio commented:
Why is the rest of #707 required?

raphinesse commented:
It does fix a theoretical bug but it's definitely not required.

For the record: I'd still prefer to have all the changes I originally approved together in a minor release.

brodybits commented:
> (Don't forget to update the PR description to reflect what you actually merged)

Done
